### PR TITLE
Use zephyr native ringbuffers

### PR DESCRIPTION
**Pull request by simensrostad**

### CODE REVIEW

File: applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c [hunk 1]
```diff
@@ -63,9 +64,56 @@ LOG_MODULE_REGISTER(cloud_codec, CONFIG_CLOUD_CODEC_LOG_LEVEL);
...
+RING_BUF_ITEM_DECLARE_SIZE(battery_buf,
+			   sizeof(struct cloud_data_battery) *
+				  CONFIG_BATTERY_BUFFER_MAX);
+
+RING_BUF_ITEM_DECLARE_SIZE(gps_buf,
+			   sizeof(struct cloud_data_gps) *
+				  CONFIG_GPS_BUFFER_MAX);
+
+/* Structure used to keep track of last known cloud codec data. */
+struct cloud_data_last_known {
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c]: `static`

File: applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c [hunk 2]
```diff
@@ -570,33 +560,81 @@ int cloud_codec_encode_data(struct cloud_codec_data *output,
...
 	}
 
-	if (bat_buf->queued) {
-		err += bat_data_add(rep_obj, bat_buf, false);
+	/* UI */
+	if (last_known.ui_valid) {
+		err = ui_data_add(rep_obj, &last_known.ui, false);
+		if (err) {
+			LOG_ERR("ui_data_add, error: %d", err);
+			return err;
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c]: This and similar returns in the function look like possible memory leaks

File: applications/asset_tracker_v2/src/cloud/cloud_codec/Kconfig [hunk 3]
```diff
@@ -4,6 +4,34 @@
 # SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
 #
 
+config GPS_BUFFER_MAX
+	int "GPS data ringbuffer entries"
+	default 15
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/Kconfig]: Buffer access is faster if the item count is a power of 2

File: applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c [hunk 4]
```diff
@@ -642,13 +687,14 @@ int cloud_codec_encode_ui_data(struct cloud_codec_data *output,
 		return -ENOMEM;
 	}
 
-	if (ui_buf->queued) {
-		err += ui_data_add(root_obj, ui_buf, false);
+	if (last_known.ui_valid) {
+		err = ui_data_add(root_obj, &last_known.ui, false);
+		if (err) {
+			LOG_ERR("ui_data_add, error: %d", err);
+			return err;
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c]: Memory leak

File: applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c [hunk 5]
```diff
@@ -689,74 +724,191 @@ int cloud_codec_encode_batch_data(
...
-			err += gps_data_add(gps_obj, &gps_buf[i], true);
+	/* UI */
+	while (ring_buf_is_empty(&ui_buf) == 0) {
+		struct cloud_data_ui *ui;
+
+		uint32_t len = ring_buf_get_claim(&ui_buf,
+						  (uint8_t **)&ui,
+						  sizeof(struct cloud_data_ui));
+		if (len != sizeof(struct cloud_data_ui)) {
+			return -ENOMEM;
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c]: memory leak

File: applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c [hunk 6]
```diff
@@ -689,74 +724,191 @@ int cloud_codec_encode_batch_data(
...
 
-	/* GPS data */
-	for (int i = 0; i < gps_buf_count; i++) {
-		if (gps_buf[i].queued) {
-			err += gps_data_add(gps_obj, &gps_buf[i], true);
+	/* UI */
+	while (ring_buf_is_empty(&ui_buf) == 0) {
+		struct cloud_data_ui *ui;
+
+		uint32_t len = ring_buf_get_claim(&ui_buf,
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c:748]: The docs are a little ambiguous on whether this API can be used on a ring buffer in item mode
```
 * @warning
 * Ring buffer instance should not mix byte access and item access
 * (calls prefixed with ring_buf_item_).
 ````

Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c:748]: We are using byte mode, where the documentation indicates that all the API functions for getting and putting data is valid.

File: applications/asset_tracker_v2/src/cloud/cloud_codec/Kconfig (hunk 3 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/Kconfig]: Im not sure I quite understood the sentence here.

Comment (jtguggedal) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/Kconfig]: Oops, typo, corrected now..

File: applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c (hunk 2 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c]: Yes, I am aware. I could clean up the error handling in the library to always cleanup before return but these errors are rather severe and any error expect -ENODATA should be rebooted on.

File: applications/asset_tracker_v2/src/cloud/cloud_codec/Kconfig (hunk 3 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/Kconfig]: Ah, ok. I see. I can add a comment and correct the entries to 16.

File: applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c (hunk 6 above)
Comment (jtguggedal) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c:748]: Oh, I guess I was confused by the macro defining the buffers: `RING_BUF_ITEM_DECLARE_SIZE`

**Comment by simensrostad** [comment]
@jtguggedal Addressed your comments. The error handling in the library should be slightly updated now.

**Comment by simensrostad** [comment]
@jtguggedal Force pushed the wrong branch?

**Comment by jtguggedal** [comment]
> @jtguggedal Force pushed the wrong branch?

I squashed doc commits into asset-tracker-v2 branch, so the branch this PR is based will need a rebase

**Comment by simensrostad** [comment]
PR will be created against master when ```asset_tracker_v2``` is merged.

**Pull request closed without merging**